src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp

```diff
@@ -0,0 +1,731 @@
+/**
+ * @file spill_tree_impl.hpp
+ *
+ * Implementation of generalized hybrid spill tree (SpillTree).
+ */
+#ifndef MLPACK_CORE_TREE_SPILL_TREE_SPILL_TREE_IMPL_HPP
+#define MLPACK_CORE_TREE_SPILL_TREE_SPILL_TREE_IMPL_HPP
+
+// In case it wasn't included already for some reason.
+#include "spill_tree.hpp"
+
+namespace mlpack {
+namespace tree {
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+SpillTree(
+    const MatType& data,
+    const double tau,
+    const size_t maxLeafSize,
+    const double rho) :
+    left(NULL),
+    right(NULL),
+    parent(NULL),
+    count(0),
+    pointsIndex(NULL),
+    overlappingNode(false),
+    hyperplane(),
+    bound(data.n_rows),
+    parentDistance(0), // Parent distance for the root is 0: it has no parent.
+    dataset(&data),
+    localDataset(false)
+{
+  std::vector<size_t> points;
+  points.reserve(dataset->n_cols);
+  for (size_t i = 0; i < dataset->n_cols; i++)
+    points.push_back(i);
```

Comment:
I think that it is better to try and use Armadillo objects here.  In general I have found they are faster (even if you are reserving space for the vector).